libvast/src/detail/sigma.cpp

```diff
@@ -0,0 +1,369 @@
+struct detection_parser : parser<detection_parser> {
+  using attribute = expression;
+
+  explicit detection_parser(const expression_map& exprs) : search_id{exprs} {
+  }
+
+  static expression to_expr(
+    std::tuple<expression, std::vector<std::tuple<bool_operator, expression>>>
+      expr) {
+    auto& [x, xs] = expr;
+    if (xs.empty())
+      return x;
+    // We split the expression chain at each OR node in order to take care of
+    // operator precedance: AND binds stronger than OR.
+    disjunction dis;
+    auto con = conjunction{x};
+    for (auto& [op, expr] : xs)
+      if (op == bool_operator::logical_and) {
+        con.emplace_back(std::move(expr));
+      } else if (op == bool_operator::logical_or) {
+        VAST_ASSERT(!con.empty());
+        if (con.size() == 1)
+          dis.emplace_back(std::move(con[0]));
+        else
+          dis.emplace_back(std::move(con));
+        con = conjunction{std::move(expr)};
+      } else {
+        VAST_ASSERT(!"negations must not exist here");
+      }
+    if (con.size() == 1)
+      dis.emplace_back(std::move(con[0]));
+    else
+      dis.emplace_back(std::move(con));
+    return dis.size() == 1 ? std::move(dis[0]) : expression{dis};
+  };
+
+  template <class Iterator>
+  bool parse(Iterator& f, const Iterator& l, expression& result) const {
+    using namespace parser_literals;
+    auto ws = ignore(*parsers::space);
+    auto negate = [](expression x) { return negation{std::move(x)}; };
+    rule<Iterator, expression> expr;
+    rule<Iterator, expression> group;
+    // clang-format off
+    group
+      = '(' >> ws >> ref(expr) >> ws >> ')'
+      | "not"_p >> ws >> '(' >> ws >> (ref(expr) ->* negate) >> ws >> ')'
+      | "not"_p >> ws >> search_id ->* negate
+      | search_id
+      ;
+    auto and_or
+      = "or"_p  ->* [] { return bool_operator::logical_or; }
+      | "and"_p  ->* [] { return bool_operator::logical_and; }
+      ;
+    expr
+      = (group >> *(ws >> and_or >> ws >> ref(group)) >> ws) ->* to_expr
+      ;
+    // clang-format on
+    auto p = expr >> parsers::eoi;
+    return p(f, l, result);
+  }
+
+  search_id_symbol_table search_id;
+};
+
+// The following invariants apply according to the Sigma spec:
+// - All values are treated as case-insensitive strings
+// - You can use wildcard characters '*' and '?' in strings
+// - Wildcards can be escaped with \, e.g. \*. If some wildcard after a
+//   backslash should be searched, the backslash has to be escaped: \\*.
+// - Regular expressions are case-sensitive by default
+// - You don't have to escape characters except the string quotation
+//   marks '
+caf::optional<pattern> make_pattern(std::string_view str) {
+  auto f = str.begin();
+  auto l = str.end();
+  std::string rx;
+  // FIXME: this is a pretty hand-wavy approach to transforming a glob string
+  // to a valid regex. We need to revisit this once we have actual pattern
+  // support in the query language.
```

Comment:
This reads like a TODO rather than a FIXME. This also needs a follow-up story.

Reply:
Story created.